packages/flutter_tools/lib/src/run_hot.dart

```diff
@@ -537,6 +537,13 @@ class HotRunner extends ResidentRunner {
           // TODO(aam): Investigate why we are validating only first reload report,
           // which seems to be current behavior
           final Map<String, dynamic> firstReport = list.first;
+          if (firstReport == null) {
+            // This could happen if the isolate is not actually running,
+            // failed to start for some reason. For example, if main method
+            // is not found in the app.
+            retrieveFirstReloadReport.completeError('Dart VM failed to provide reload report');
```

Comment:
maybe the message should state
'Reload failed, please restart the application'

Reply:
This message comes up in context where it is clear that hot reload failed, it's purpose seem to provide more details as to why it failed:
```
Initializing hot reload...
Error 105 received from application: Isolate must be runnable
Try again after fixing the above error(s).
Hot reload failed: Dart VM failed to provide reload report
```

Reply:
In that case would it make sense to use the actual error
'main' not found
in this case.

The message 'Dart VM failed to provide reload report' sounds like an internal error to me.

Reply:
I think it is kind of internal error, it's just we are handling it more gracefully. In the console output you can actually see that we complained that `main` was not found first:
```
Initializing gradle...
Built build/app/outputs/apk/debug/app-debug.apk (30.3MB).
Installing build/app/outputs/apk/app.apk...
I/FlutterActivityDelegate(18782): onResume setting current activity to this
E/flutter (18782): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Dart_LoadScriptFromKernel: The binary program does not contain 'main'.
E/flutter (18782): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Dart_GetClosure expects argument 'library' to be non-null.
Syncing files to device Android SDK built for x86...

🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
An Observatory debugger and profiler on Android SDK built for x86 is available at: http://127.0.0.1:8104/
For a more detailed help message, press "h". To quit, press "q".
r

Initializing hot reload...
Error 105 received from application: Isolate must be runnable
Try again after fixing the above error(s).
Hot reload failed: Dart VM failed to provide reload report
```

Reply:
The app did not start in the first place because it got the error 'The binary program does not contain 'main'', why did we not quit the execution instead of displaying the 'hot reload' prompt.
What does it mean to 'hot reload' an app that did not start at all ?

Reply:
This PR addresses just a crashing part of current user experience. The app still doesn't run, but flutter tools doesn't crash anymore, just produces an error.

Reply:
Ideally in this particular case we stop flutter tools from progressing forward after request to run flutter app failed, but it will take some more effort because currently there is no api for engine to report errors during launch. Also, there is engine refactor that should land first before the api is adjusted further.

Reply:
Ok, maybe a TODO here about stopping flutter tools in this case.
Can we also change the message to 'launch of app failed' or something like that.

Reply:
Added TODO.

The message is already says "Hot reload failed:". This adds some more context as to the reason why it failed, so the end result looks like this: "Hot reload failed: Dart VM failed to provide reload report".

Reply:
Turns out there is better way to handle this as `null` overrides properly formed error message in Flutter vmservice.